stdlib/src/commonMain/kotlin/it/unibo/collektive/stdlib/Spreading.kt

```diff
@@ -0,0 +1,118 @@
+ *
+ * If there are no sources and no neighbors, default to [local] value.
+ * The [metric] function is used to compute the distance between devices in form of a field of [Distance]s.
+ * [Distance]s must be in the [[bottom], [top]] range, [accumulateDistance] is used to sum distances.
+ * [accumulateData] is used to modify data from neighbors on the fly, and defaults to the identity function.
+ */
+@JvmOverloads
+inline fun <ID : Any, Value, Distance : Comparable<Distance>> Aggregate<ID>.gradientCast(
+    source: Boolean,
+    local: Value,
+    bottom: Distance,
+    top: Distance,
+    crossinline accumulateData: (Distance, Distance, Value) -> Value = { _, _, data -> data },
+    crossinline accumulateDistance: (Distance, Distance) -> Distance,
+    crossinline metric: () -> Field<ID, Distance>,
+): Value {
+    val topValue = top to local
+    return share(topValue) { neighborData ->
+        val paths = neighborData.alignedMap(metric()) { (fromSource, data), toNeighbor ->
+            val totalDistance = accumulateDistance(fromSource, toNeighbor)
+            val newData = accumulateData(fromSource, toNeighbor, data)
+            totalDistance to newData
+        }
+        when {
+            source -> bottom to local
+            else -> paths.minBy(base = topValue) { it.first } // sort by distance from the nearest source
+        }
+    }.second // return the data
+}
+
+/**
+ * Propagate [local] values across a spanning tree starting from the closest [source].
+ *
+ * If there are no sources and no neighbors, default to [local] value.
+ * The [metric] function is used to compute the distance between devices in form of a field of [Int]s.
+ * [accumulateData] is used to modify data from neighbors on the fly, and defaults to the identity function.
+ */
+@JvmOverloads
+@JvmName("gradientCastInt")
+inline fun <ID : Any, Type> Aggregate<ID>.gradientCast(
+    source: Boolean,
+    local: Type,
+    crossinline accumulateData: (Int, Int, Type) -> Type = { _, _, data -> data },
+    crossinline metric: () -> Field<ID, Int> = { neighboring(1) },
+): Type = gradientCast(source, local, Int.MIN_VALUE, Int.MAX_VALUE, accumulateData, Int::plus, metric)
+
+/**
+ * Propagate [local] values across a spanning tree starting from the closest [source].
+ *
+ * If there are no sources and no neighbors, default to [local] value.
+ * The [metric] function is used to compute the distance between devices in form of a field of [Double]s.
+ * [accumulateData] is used to modify data from neighbors on the fly, and defaults to the identity function.
+ */
+@JvmOverloads
+@JvmName("gradientCastDouble")
+inline fun <ID : Any, Type> Aggregate<ID>.gradientCast(
+    source: Boolean,
+    local: Type,
+    crossinline accumulateData: (Double, Double, Type) -> Type = { _, _, data -> data },
+    crossinline metric: () -> Field<ID, Double> = { neighboring(1.0) },
+): Type = gradientCast(source, local, 0.0, Double.POSITIVE_INFINITY, accumulateData, Double::plus, metric)
```

Comment:
I cannot understand why in the `Int` version of the gradient you used `Int.MIN_VALUE` but in this version you used `0`. Is there any reason for that?

Reply:
Yes. Floats lose precision at the extremes, and using a negative value would shift the maximum precision point away from the gradient source.